daemon/src/GHCSpecter/Control.hs

```diff
@@ -0,0 +1,251 @@
+module GHCSpecter.Control
+  ( Control,
+    type Runner,
+    control,
+    stepControl,
+    stepControlUpToEvent,
+  )
+where
+
+import Concur.Core (Widget, unsafeBlockingIO)
+import Control.Lens ((.~), (^.), _1, _2)
+import Control.Monad (forever)
+import Control.Monad.Extra (loopM)
+import Control.Monad.Free (Free (..), liftF)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.State (StateT (..), get, modify', put)
+import Data.Aeson (encode)
+import Data.ByteString.Lazy qualified as BL
+import Data.IORef (IORef, modifyIORef', newIORef, readIORef)
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Text.IO qualified as TIO
+import Data.Time.Clock (NominalDiffTime, UTCTime)
+import Data.Time.Clock qualified as Clock
+import GHCSpecter.Channel (SessionInfo (..))
+import GHCSpecter.Server.Types
+  ( HasServerState (..),
+    ServerState (..),
+  )
+import GHCSpecter.UI.ConcurReplica.Types (IHTML)
+import GHCSpecter.UI.Types
+  ( HasModuleGraphUI (..),
+    HasSourceViewUI (..),
+    HasTimingUI (..),
+    HasUIState (..),
+    ModuleGraphUI (..),
+    UIState (..),
+  )
+import GHCSpecter.UI.Types.Event
+  ( Event (..),
+    ModuleGraphEvent (..),
+    SessionEvent (..),
+    SubModuleEvent (..),
+    TimingEvent (..),
+  )
+import System.IO (IOMode (..), withFile)
+import System.IO.Unsafe (unsafePerformIO)
+
+data ControlF r
+  = GetState ((UIState, ServerState) -> r)
+  | PutState (UIState, ServerState) r
+  | NextEvent (Event -> r)
+  | PrintMsg Text r
+  | GetCurrentTime (UTCTime -> r)
+  | GetLastUpdatedUI (UTCTime -> r)
+  | ShouldUpdate Bool r
+  | SaveSession r
+  deriving (Functor)
+
+type Control = Free ControlF
+
+getState :: Control (UIState, ServerState)
+getState = liftF (GetState id)
+
+putState :: (UIState, ServerState) -> Control ()
+putState (ui, ss) = liftF (PutState (ui, ss) ())
+
+nextEvent :: Control Event
+nextEvent = liftF (NextEvent id)
+
+printMsg :: Text -> Control ()
+printMsg txt = liftF (PrintMsg txt ())
+
+getCurrentTime :: Control UTCTime
+getCurrentTime = liftF (GetCurrentTime id)
+
+getLastUpdatedUI :: Control UTCTime
+getLastUpdatedUI = liftF (GetLastUpdatedUI id)
+
+shouldUpdate :: Bool -> Control ()
+shouldUpdate b = liftF (ShouldUpdate b ())
+
+saveSession :: Control ()
+saveSession = liftF (SaveSession ())
+
+-- TODO: remove this
+tempRef :: IORef Int
+tempRef = unsafePerformIO (newIORef 0)
+{-# NOINLINE tempRef #-}
+
+type Runner = StateT (UIState, ServerState) (Widget IHTML)
+
+-- | step interpretation
```

Comment:
this interpretation is more-primitive per-pattern interpretation, regardless of whether the interpreter yields its operation to UI, than up-to-event-handling interpretation. 
`stepControlUpToEvent` is the per-event-handling interpretation.

Reply:
Added a sectional Note [Control Loops].